File: app/src/main/java/Day23.java
import shared.Graph;

import java.util.*;

public class Day23 {
    public static <T> List<T> doBreadthFirstSearch(T start, Graph<T> graph) {
         //set of visited vertex in the order they were found
        Set<T> visited = new LinkedHashSet<>();
        Queue<T> queue = new ArrayDeque<>();
        queue.add(start);
        visited.add(start);

        while(!queue.isEmpty()) {
            //pull the item from the top of the list
            T currVertex = queue.poll();
            List<T> adjacencyList = graph.getAdjList(currVertex);

            for(T adj : adjacencyList) {
                //if we haven't already visited the vertex, add to both
                if(visited.add(adj)) {
                    queue.add(adj);
                }
            }
        }

        return new ArrayList<>(visited);
    }
}

File: app/src/main/kotlin/Day23.kt
import shared.Graph

fun <T> doBreadthFirstSearch(start: T, graph: Graph<T>): List<T> {
    val visited = linkedSetOf<T>().apply { add(start) }
    val queue = ArrayDeque<T>().apply { add(start) }

    while (queue.isNotEmpty()) {
        val curr = queue.removeFirst()
        graph.getAdjList(curr).forEach { adj ->
            if (visited.add(adj)) queue += adj
        }
    }

    return visited.toList()
}